File: Runtime/Samples/sympl/python/lexer.py


class Lexer (object):
    def __init__ (self, reader):
        self.Reader = reader
        self._putToken = None

    def PutToken (self, token):
        debugprint("puttoken: " + str(token))
        if self._putToken is not None:
            error("Internal Error: putting token when there is one?")
        self._putToken = token

    _numChars = [chr(x) for x in xrange(ord('0'), ord('9') + 1)]

    ### GetToken is one of two main entry points to Lexer.
    ###
    def GetToken (self):
        token = self._getToken()
        debugprint("getoken: " + str(token))
        return token

    ### Returns any saved _putToken, else skips whitespace and returns next
    ### token from input stream.
    ###
    ### If returning token directly based on char, need to gobble char, but if
    ### calling helper function to read more, then they gobble as needed.
    ###
    ### Need to maintain src loc info and store in token.
    ###
    def _getToken (self):
        if self._putToken is not None:
            tmp = self._putToken
            self._putToken = None
            debugprint("returning put token" + str(tmp))
            return tmp
        self._skipWhitespace()
        ch = self._peekChar()
        debugprint("gettoken: peek char is " + 
                   ((ch is SyntaxToken.EOF and "EOF") or ch))
        if ch is SyntaxToken.EOF:
            return ch
        elif ch == '(':
            self._getChar()
            return SyntaxToken.Paren
        elif ch == ')':
            self._getChar()
            return SyntaxToken.CloseParen
        elif ch in self._numChars:
            return self._getNumber()
        elif ch == '"':
            return self._getString()
        elif ch == "'":
            self._getChar()
            return SyntaxToken.Quote
        elif ch == '-':
            return self._getIdOrNumber()
        elif self._startsId(ch):
            return self._getIdOrKeyword()
        elif ch == '.':
            self._getChar()
            return SyntaxToken.Dot
        raise Exception("Internal: couldn't get token? -- char[" + 
                        str(ch) + "]")

    ### _getIdOrNumber expects a hyphen as the current char, and returns an Id
    ### or number token.
    ###
    def _getIdOrNumber (self):
        ch = self._getChar()
        if ch is SyntaxToken.EOF or ch != '-':
            raise Exception("Internal: parsing ID or number without hyphen.")
        ch = self._peekChar()
        if ch in self._numChars:
            token = self._getNumber()
            token.Value = - token.Value
            return token
        elif not self._isIdTerminator(ch):
            return self._getIdOrKeyword("-")
        else:
            return self._makeIdOrKeywordToken("-")

    ### GetIdOrKeyword has first param to handle call from GetIdOrNumber where
    ### ID started with a hyphen (and looked like a number).  Didn't want to
    ### add hyphen to StartId test since it normally appears as the keyword
    ### minus.
    ###
    ### Must not call when the next char is EOF.
    ###
    def _getIdOrKeyword (self, first = None):
        ## Process first and set initial result.
        quotedId = False
        first = first or self._getChar()
        if first is SyntaxToken.EOF or not self._startsId(first):
            raise Exception("Internal: getting Id or keyword?")
        if first == '\\':
            quotedId = True
            res = self._getChar()
            if res is SyntaxToken.EOF:
                raise Exception("Unexpected EOF when getting Id.")
            if not self._startsId(first):
                raise Exception("Don't support quoted Ids that have " +
                                "non Id constituent characters.")
        else:
            res = first
        ## See if there's more chars to Id
        c = self._peekChar()
        while (c is not SyntaxToken.EOF) and (not self._isIdTerminator(c)):
            res = res + c
            self._getChar()
            c = self._peekChar()
        return self._makeIdOrKeywordToken(res, quotedId)

    ### Keep whatever casing we found in the source program so that when the
    ### IDs are used a member names and metadata on binders, then if some MO
    ### doesn't respect the IgnoreCase flag, there's an out for still binding.
    ###
    def _makeIdOrKeywordToken (self, name, quotedId = False):
        if not quotedId and KeywordToken.IsKeywordName(name):
            return KeywordToken.GetKeywordToken(name)
        else:
            if name.lower() == "let":
                print "\nWARNING: using 'let'?  You probaby meant let*.\n"
            return IdOrKeywordToken(name, False)
    
    ### Must not be called on SyntaxToken.EOF
    def _startsId (self, c):
        return c == '\\' or not self._isIdTerminator(c)

    ## Restrict macro syntax chars in case try to add macros later, but need to
    ## allow backquote in IDs to support .NET type names that come from
    ## reflection.  We can fix this later with more machinery around type names.
    _id_terminators = ['(', ')', '"', ";", ',', '@', "'", "."] #'`', 
    
    def _isIdTerminator (self, c):
        return (c in self._id_terminators) or (ord(c) < 33)


    ### _getNumber returns parsed integers as NumberTokens.  Need to update
    ### and use .NET's System.Double.Parse after scanning to non-constituent
    ### char.
    ###
    def _getNumber (self):
        ## Check integrity before loop to avoid incidentally returning zero.
        c = self._getChar()
        if c is SyntaxToken.EOF or c not in self._numChars:
            raise Exception("Internal: lexing number?")
        digit = ord(c) - ord('0')
        res = digit
        c = self._peekChar()
        while (c is not SyntaxToken.EOF and c in self._numChars):
            res = (res * 10) + (ord(c) - ord('0'))
            self._getChar()
            c = self._peekChar()
        return NumberToken(res)

    def _getString (self):
        c = self._getChar()
        if c is SyntaxToken.EOF or c != '"':
            raise Exception("Internal: parsing string?")
        res = ""
        escape = False
        c = self._peekChar()
        while True:
            if c is SyntaxToken.EOF:
                raise Exception("Hit EOF in string literal.")
            elif c == '\n' or c == '\r':
                raise Exception("Hit newline in string literal")
            elif c == '\\' and not escape:
                self._getChar()
                escape = True
            elif c == '"' and not escape:
                self._getChar()
                return StringToken(res)
            elif escape:
                escape = False
                self._getChar()
                if c == 'n':
                    res = res + "\n"
                elif c == 't':
                    res = res + "\t"
                elif c == 'r':
                    res = res = "\r"
                elif c == '"':
                    res = res + c
                elif c == '\\':
                    res = res + c
            else:
                self._getChar()
                res = res + c
            c = self._peekChar()

    _whitespaceChars = [' ', '\r', '\n', ';', '\t']

    def _skipWhitespace (self):
        ch = self._peekChar()
        while ch in self._whitespaceChars:
            debugprint("gobble: " + ch)
            if ch == ';':
                self._getChar()
                ch = self._peekChar()
                ## If newline seq is two chars, second gets eaten in outer loop.
                while ch != '\n' and ch != '\r':
                    if ch is SyntaxToken.EOF: return
                    self._getChar()
                    ch = self._peekChar()
            else:
                self._getChar()
                ch = self._peekChar()

    def _getChar (self):
        c = self.Reader.Read()
        if c == -1: return SyntaxToken.EOF
        #debugprint("_getChar: " + chr(c))
        return chr(c)

    def _peekChar (self):
        c = self.Reader.Peek()
        if c == -1: return SyntaxToken.EOF
        #debugprint("_peekChar: " + chr(c))
        return chr(c)


class Token (object):
    def __init__ (self, srcloc):
        self.SrcLoc = srcloc
    pass

class LiteralToken (Token):
    def __init__ (self, val, srcloc = (0, 0)):
        Token.__init__(self, srcloc)
        self.Value = val

class NumberToken (LiteralToken):
    def __init__ (self, num, srcloc = (0, 0)):
        LiteralToken.__init__(self, num, srcloc)

class StringToken (LiteralToken):
    def __init__ (self, string, srcloc = (0, 0)):
        LiteralToken.__init__(self, string, srcloc)

### IdOrKeywordToken represents identifier.  A subtype, KeywordToken, reps
### keywords.  The parser handles when keywords can be used like identifiers,
### for example, as .NET members when importing and renaming, literal kwd
### constants (nil, true, false), etc.  These are used also when parsing list
### literals before they get converted to runtime Symbol types by etgen.
###
class IdOrKeywordToken (Token):
    def __init__ (self, name, kwd, srcloc = (0, 0)):
        Token.__init__(self, srcloc)
        self.Name = name
        self.IsKeywordToken = kwd
    def __repr__ (self):
        if self.IsKeywordToken:
            return "<Kwd: " + self.Name + ">"
        else:
            return "<Id: " + self.Name + ">"


class KeywordToken (IdOrKeywordToken):
    def __init__ (self, name, srcloc = (0, 0)):
        IdOrKeywordToken.__init__(self, name, True, srcloc)
        self.Name = name
        KeywordToken._keywords[name.lower()] = self
    
    _keywords = {}
    
    @staticmethod
    def IsKeywordName (name):
        return name.lower() in KeywordToken._keywords
    
    @staticmethod
    def GetKeywordToken (name):
        return KeywordToken._keywords[name.lower()]

KeywordToken.Import = KeywordToken("Import")
KeywordToken.Defun = KeywordToken('Defun')
KeywordToken.Lambda = KeywordToken('Lambda')
KeywordToken.Defclass = KeywordToken('DefClass')
KeywordToken.Defmethod = KeywordToken('DefMethod')
KeywordToken.New = KeywordToken('New')
KeywordToken.Set = KeywordToken('Set')
KeywordToken.LetStar = KeywordToken('Let*')
KeywordToken.Block = KeywordToken('Block')
KeywordToken.Loop = KeywordToken('Loop')
KeywordToken.Break = KeywordToken('Break')
KeywordToken.Continue = KeywordToken('Continue')
KeywordToken.Return = KeywordToken('Return')
KeywordToken.List = KeywordToken('List')
KeywordToken.Cons = KeywordToken('Cons')
KeywordToken.Eq = KeywordToken('Eq')
KeywordToken.Elt = KeywordToken('Elt')
KeywordToken.Nil = KeywordToken('Nil')
KeywordToken.True = KeywordToken('True')
KeywordToken.False = KeywordToken('False')
KeywordToken.If = KeywordToken('If')
KeywordToken.Add = KeywordToken('+')
KeywordToken.Subtract = KeywordToken('-')
KeywordToken.Multiply = KeywordToken('*')
KeywordToken.Divide = KeywordToken('/')
KeywordToken.Equal = KeywordToken('=')
KeywordToken.NotEqual = KeywordToken('!=')
KeywordToken.GreaterThan = KeywordToken('>')
KeywordToken.LessThan = KeywordToken('<')
KeywordToken.And = KeywordToken('And')
KeywordToken.Or = KeywordToken('Or')
KeywordToken.Not = KeywordToken('Not')



class SyntaxToken (Token):
    def __init__ (self, name):
        self.Name = name
    def __str__ (self): return self.Name
    def __repr__ (self): return "<SyntaxToken " + self.Name + ">"
SyntaxToken.Paren = SyntaxToken('Paren')
SyntaxToken.CloseParen = SyntaxToken('CloseParen')
SyntaxToken.EOF = SyntaxToken('EOF')
SyntaxToken.Quote = SyntaxToken('Quote')
SyntaxToken.Dot = SyntaxToken('Dot')


##################
### Dev-time Utils
##################

_debug = False
def debugprint (*stuff):
    if _debug:
        for x in stuff:
            print x,
        print
_debug = False

File: Runtime/Samples/sympl/csharp/Lexer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace SymplSample {

    public class Lexer {
        private Token _token = null;
        private TextReader _reader;
        private const char _eof = unchecked ((char)(-1));

        public Lexer(TextReader reader) {
            if (reader == null) {
                throw new ArgumentException("reader");
            }
            _reader = reader;
        }

        public void PutToken(Token token) {
            if (_token != null) {
                throw new InvalidOperationException(
                    "Internal Error: putting token when there is one?");
            } else {
                _token = token;
            }
        }

        // Returns any saved _putToken, else skips whitespace and returns next
        // token from input stream.
        //
        // If returning token directly based on char, need to gobble char, but
        // if calling helper function to read more, then they gobble as needed.
        //
        // TODO: maintain src loc info and store in token.
        //
        public Token GetToken() {
            if (_token != null) {
                Token tmp = _token;
                _token = null;
                return tmp;
            }
            SkipWhitespace();
            char ch = PeekChar();
            if (ch == _eof) {
                return SyntaxToken.EOF;
            } else if (ch == '(') {
                GetChar();
                return SyntaxToken.Paren;
            } else if (ch == ')') {
                GetChar();
                return SyntaxToken.CloseParen;
            } else if (IsNumChar(ch)) {
                return GetNumber();
            } else if (ch == '\"') {
                return GetString();
            } else if (ch == '\'') {
                GetChar();
                return SyntaxToken.Quote;
            } else if (ch == '-') {
                return GetIdOrNumber();
            } else if (StartsId(ch)) {
                return GetIdOrKeyword();
            } else if (ch == '.') {
                GetChar();
                return SyntaxToken.Dot;
            }
            throw new InvalidOperationException("Internal: couldn't get token?");
        }

        // _getIdOrNumber expects a hyphen as the current char, and returns an Id
        // or number token.
        //
        private Token GetIdOrNumber() {
            char ch = GetChar();
            if (ch == _eof || ch != '-') {
                throw new InvalidOperationException(
                    "Internal: parsing ID or number without hyphen.");
            }
            ch = PeekChar();
            if (IsNumChar(ch)) {
                NumberToken token = GetNumber();
                return new NumberToken(-(int)(token.Value));
            } else if (!IsIdTerminator(ch)) {
                return GetIdOrKeyword('-');
            } else {
                return MakeIdOrKeywordToken("-", false);
            }
        }

        // GetIdOrKeyword has first param to handle call from GetIdOrNumber where
        // ID started with a hyphen (and looked like a number).  Didn't want to
        // add hyphen to StartId test since it normally appears as the keyword
        // minus.  Usually the overload without the first param is called.
        //
        // Must not call when the next char is EOF.
        //
        private IdOrKeywordToken GetIdOrKeyword() {
            return GetIdOrKeyword(GetChar());
        }
        private IdOrKeywordToken GetIdOrKeyword(char first) {
            bool quotedId = false;
            if (first == _eof || !StartsId(first)) {
                throw new InvalidOperationException(
                    "Internal: getting Id or keyword?");
            }
            StringBuilder res = new StringBuilder();
            char c;
            if (first == '\\') {
                quotedId = true;
                c = GetChar();
                if (c == _eof) {
                    throw new InvalidOperationException(
                        "Unexpected EOF when getting Id.");
                }
                if (!StartsId(first)) {
                    throw new InvalidOperationException(
                        "Don't support quoted Ids that have non " + 
                        "Id constituent characters.");
                }
                res.Append(c);
            } else {
                res.Append(first);
            }
            // See if there's more chars to Id
            c = PeekChar();
            while (c != _eof && (!IsIdTerminator(c))) {
                res.Append(c);
                GetChar();
                c = PeekChar();
            }
            return MakeIdOrKeywordToken(res.ToString(), quotedId);
        }

        // Keep whatever casing we found in the source program so that when the
        // IDs are used a member names and metadata on binders, then if some MO
        // doesn't respect the IgnoreCase flag, there's an out for still binding.
        //
        private IdOrKeywordToken MakeIdOrKeywordToken(string name, bool quotedId) {
            if (!quotedId && KeywordToken.IsKeywordName(name)) {
                return KeywordToken.GetKeywordToken(name);
            } else {
                if (name.ToLower() == "let") {
                    System.Console.WriteLine();
                    System.Console.WriteLine(
                       "WARNING: using 'let'?  You probably meant let*.");
                    System.Console.WriteLine();
                }
                return new IdOrKeywordToken(name);
            }
        }

        // Must not be called on EOF
        //
        private bool StartsId(char c) {
            return c == '\\' || !IsIdTerminator(c);
        }

        // Restrict macro syntax chars in case try to add macros later, but need to
        // allow backquote in IDs to support .NET type names that come from
        // reflection.  We can fix this later with more machinery around type names.
        private static char[] _id_terminators = {'(', ')', '\"', ';', ',', /*'`',*/
                                                 '@', '\'', '.'};
        private static bool IsIdTerminator(char c) {
            return _id_terminators.Contains(c) || (c < (char)33);
        }

        // _getNumber returns parsed integers as NumberTokens.  Need to update
        // and use .NET's System.Double.Parse after scanning to non-constituent
        // char.
        //
        private NumberToken GetNumber() {
            // Check integrity before loop to avoid accidently returning zero.
            char c = GetChar();
            if (c == _eof || !IsNumChar((char)c)) {
                throw new InvalidOperationException("Internal: lexing number?");
            }
            int digit = c - '0';
            int res = digit;
            c = PeekChar();
            while (c != _eof && IsNumChar(c)) {
                res = res * 10 + (c - '0');
                GetChar();
                c = PeekChar();
            }
            return new NumberToken(res);
        }

        private StringToken GetString() {
            char c = GetChar();
            if (c == _eof || c != '\"') {
                throw new InvalidOperationException(
                    "Internal: parsing string?");
            }
            StringBuilder res = new StringBuilder();
            bool escape = false;
            c = PeekChar();
            while (true) {
                if (c == _eof) {
                    throw new InvalidOperationException(
                        "Hit EOF in string literal.");
                } else if (c == '\n' || c == '\r') {
                    throw new InvalidOperationException(
                        "Hit newline in string literal");
                } else if (c == '\\' && !escape) {
                    GetChar();
                    escape = true;
                } else if (c == '"' && !escape) {
                    GetChar();
                    return new StringToken(res.ToString());
                } else if (escape) {
                    escape = false;
                    GetChar();
                    switch (c) {
                        case 'n':
                            res.Append('\n');
                            break;
                        case 't':
                            res.Append('\t');
                            break;
                        case 'r':
                            res.Append('\r');
                            break;
                        case '\"':
                            res.Append('\"');
                            break;
                        case '\\':
                            res.Append('\\');
                            break;
                    }
                } else {
                    GetChar();
                    res.Append(c);
                }
                c = PeekChar();
            }
        } //GetString
        
        private static char[] _whitespaceChars = { ' ', '\r', '\n', ';', '\t' };

        private void SkipWhitespace() {
            char ch = PeekChar();
            while (_whitespaceChars.Contains((char)ch)) {
                if (ch == ';') {
                    do {
                        GetChar();
                        ch = PeekChar();
                        if (ch == _eof) return;
                    // If newline seq is two chars, second gets eaten in outer loop.
                    } while (ch != '\n' && (char)ch != '\r');
                } else {
                    GetChar();
                    ch = PeekChar();
                }
            }
        }

        private char GetChar() {
            return unchecked ((char)_reader.Read());
        }

        private char PeekChar() {
            return unchecked ((char)_reader.Peek());
        }

        private static bool IsNumChar(char c) {
            return c >= '0' && c <= '9';
        }
    } //Lexer


    public abstract class Token {
        // TODO: Add source location to token
    }

    internal class LiteralToken : Token {
        private object _value;

        public object Value {
            get { return _value; }
        }

        public LiteralToken(object val) {
            _value = val;
        }
    }

    internal class NumberToken : LiteralToken {
        public NumberToken(int val) 
            : base(val)
        {
        }
    }

    internal class StringToken : LiteralToken {
        public StringToken(string str)
            : base(str) {
        }
    }


    // IdOrKeywordToken represents identifier.  A subtype, KeywordToken, reps
    // keywords.  The parser handles when keywords can be used like identifiers,
    // for example, as .NET members when importing and renaming, literal kwd
    // constants (nil, true, false), etc.  These are used also when parsing list
    // literals before they get converted to runtime Symbol types by etgen.
    //
    public class IdOrKeywordToken : Token {
        private string _name;

        public IdOrKeywordToken(string id) {
            this._name = id;
        }

        public string Name { get { return _name; } }

        public virtual bool IsKeywordToken { get { return false; } }
    }


    internal class KeywordToken : IdOrKeywordToken {

        static KeywordToken() {
            InitializeKeywords();
        }

        private KeywordToken(string id) 
            : base(id) {
        }

        private static Dictionary<string, KeywordToken> _keywords =
            new Dictionary<string,KeywordToken>();

        private static void InitializeKeywords() {
            _keywords["import"] = Import;
            _keywords["defun"] = Defun;
            _keywords["lambda"] = Lambda;
            _keywords["defclass"] = Defclass;
            _keywords["defmethod"] = Defmethod;
            _keywords["new"] = New;
            _keywords["set"] = Set;
            _keywords["let*"] = LetStar;
            _keywords["block"] = Block;
            _keywords["loop"] = Loop;
            _keywords["break"] = Break;
            _keywords["continue"] = Continue;
            _keywords["return"] = Return;
            _keywords["cons"] = Cons;
            _keywords["eq"] = Eq;
            _keywords["list"] = List;
            _keywords["elt"] = Elt;
            _keywords["nil"] = Nil;
            _keywords["true"] = True;
            _keywords["if"] = If;
            _keywords["false"] = False;
            _keywords["+"] = Add;
            _keywords["-"] = Substract;
            _keywords["*"] = Muliply;
            _keywords["/"] = Divide;
            _keywords["="] = Equal;
            _keywords["!="] = NotEqual;
            _keywords[">"] = GreaterThan;
            _keywords["<"] = LessThan;
            _keywords["and"] = And;
            _keywords["or"] = Or;
            _keywords["not"] = Not;
        }

        public static KeywordToken Import = new KeywordToken("Import");
        public static KeywordToken Defun = new KeywordToken("Defun");
        public static KeywordToken Lambda = new KeywordToken("Lambda");
        public static KeywordToken Defclass = new KeywordToken("Defclass");
        public static KeywordToken Defmethod = new KeywordToken("Defmethod");
        public static KeywordToken New = new KeywordToken("New");
        public static KeywordToken Set = new KeywordToken("Set");
        public static KeywordToken LetStar = new KeywordToken("LetStar");
        public static KeywordToken Block = new KeywordToken("Block");
        public static KeywordToken Loop = new KeywordToken("Loop");
        public static KeywordToken Break = new KeywordToken("Break");
        public static KeywordToken Continue = new KeywordToken("Continue");
        public static KeywordToken Return = new KeywordToken("Return");
        public static KeywordToken List = new KeywordToken("List");
        public static KeywordToken Cons = new KeywordToken("Cons");
        public static KeywordToken Eq = new KeywordToken("Eq");
        public static KeywordToken Elt = new KeywordToken("Elt");
        public static KeywordToken Nil = new KeywordToken("Nil");
        public static KeywordToken True = new KeywordToken("True");
        public static KeywordToken If = new KeywordToken("If");
        public static KeywordToken False = new KeywordToken("False");
        public static KeywordToken Add = new KeywordToken("+");
        public static KeywordToken Substract = new KeywordToken("-");
        public static KeywordToken Muliply = new KeywordToken("*");
        public static KeywordToken Divide = new KeywordToken("/");
        public static KeywordToken Equal = new KeywordToken("=");
        public static KeywordToken NotEqual = new KeywordToken("!=");
        public static KeywordToken GreaterThan = new KeywordToken(">");
        public static KeywordToken LessThan = new KeywordToken("<");
        public static KeywordToken And = new KeywordToken("And");
        public static KeywordToken Or = new KeywordToken("Or");
        public static KeywordToken Not = new KeywordToken("Not");

        public override bool IsKeywordToken {
            get {return true;}
        }

        internal static KeywordToken GetKeywordToken(string name) {
            return _keywords[name.ToLower()];
        }

        internal static bool IsKeywordName(string id) {
            return _keywords.ContainsKey(id.ToLower());
        }
    } //KeywordToken


    // SyntaxTokenKind is used for debugging.  The parser does identity check on
    // SyntaxToken members.
    //
	internal enum SyntaxTokenKind {
        Paren,
        CloseParen,
        EOF,
        Quote,
        Dot,
    }

    internal class SyntaxToken : Token {
        private SyntaxTokenKind _kind;

        private SyntaxToken(SyntaxTokenKind kind) {
            _kind = kind;
        }

        public override string ToString() {
            return "<SyntaxToken " + _kind.ToString() + ">";
        }

        public static SyntaxToken Paren = new SyntaxToken(SyntaxTokenKind.Paren);
        public static SyntaxToken CloseParen = 
            new SyntaxToken(SyntaxTokenKind.CloseParen);
        public static SyntaxToken EOF = new SyntaxToken(SyntaxTokenKind.EOF);
        public static SyntaxToken Quote = new SyntaxToken(SyntaxTokenKind.Quote);
        public static SyntaxToken Dot = new SyntaxToken(SyntaxTokenKind.Dot);
    }
}
